addons/pkg-maven/common/src/main/java/org/commonjava/indy/pkg/maven/content/MavenMetadataGenerator.java

```diff
@@ -368,65 +370,86 @@ public Transfer generateGroupFileContent( final Group group, final List<Artifact
+                        new MetadataXpp3Writer().write( baos, md );
+
+                        final byte[] merged = baos.toByteArray();
+                        if ( merged != null )
                         {
-                            closeQuietly( fos );
-                        }
+                            OutputStream fos = null;
+                            try
+                            {
+                                fos = target.openOutputStream( TransferOperation.GENERATE, true, eventMetadata );
+                                fos.write( merged );
+                            }
+                            catch ( final IOException e )
+                            {
+                                throw new IndyWorkflowException( "Failed to write merged metadata to: {}.\nError: {}",
+                                                                 e, target, e.getMessage() );
+                            }
+                            finally
+                            {
+                                closeQuietly( fos );
+                            }
 
-                        writeGroupMergeInfo( group, members, toMergePath );
+                            writeGroupMergeInfo( group, members, toMergePath );
+                        }
                     }
-                }
-                catch ( final IOException e )
-                {
-                    logger.error( String.format( "Cannot write consolidated metadata: %s to: %s. Reason: %s", path,
-                                                 group.getKey(), e.getMessage() ), e );
-                }
-
-                if ( exists( target ) )
-                {
-                    //                return target;
-                    // if this is a checksum file, we need to return the original path.
-                    Transfer original = fileManager.getTransfer( group, path );
-                    if ( exists( original ) )
+                    catch ( final IOException e )
                     {
-                        return original;
+                        logger.error( String.format( "Cannot write consolidated metadata: %s to: %s. Reason: %s", path,
+                                                     group.getKey(), e.getMessage() ), e );
                     }
                 }
             }
-
-            return null;
+            finally
+            {
+                mergerLock.unlock();
+            }
         }
-        finally
+        else
         {
-            target.unlock();
+            logger.info(
+                    "The metadata generation is still in process by another thread for the metadata file for this path {} in group {}, so block current thread to wait for result",
+                    path, group );
+            try
+            {
+                //TODO: potential dead lock if the real merging thread is blocked by some underline issues.
+                mergerLock.lock();
```

Comment:
What is this lock/unlock section for? It looks like it could hang the system without accomplishing anything?